### PR TITLE
Prevent upgrade failures if there are existing annotations permissions

### PR DESCRIPTION
**Pull request by nvazquez**
### Description

This PR prevents upgrade failures in the specific case in which the annotations are allowed as part of the role permissions for some roles.
Fixes: #5617 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?
- Create a vanilla environment
- Upgrade a 4.15.2 environment without adding any role permission

### CODE REVIEW

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]
cltgm

**Review by GutoVeronezi — APPROVED** [review]
CLGTM

**Review by GabrielBrascher — APPROVED** [review]
LGTM

**Review by rohityadavcloud — APPROVED** [review]

**Comment by rohityadavcloud** [comment]
LGTM - would it make more sense to do via java (to check if these exist?)

**Comment by nvazquez** [comment]
@rohityadavcloud sorry I do not see the benefit of moving it to the java code in this case, do you mean moving the SQL check to the java upgrade code or updating the `authorized` parameter on the @API annotation for the java classes involved?

**Comment by DaanHoogland** [comment]
@nvazquez there is a benefit in what @rohityadavcloud says; we are in a mess with guest OS mappings because of using IDs in schema upgrade scripts. This could happen for roles as well in the long run. So it might be good to device an upgrade mech that is more robust. This will work for sure, but it is not a mechanism that we should promote.

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@DaanHoogland @rohityadavcloud I see your point, thanks for clarifying. I have updated the PR accordingly, can you please re-review?

**Comment by nvazquez** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41520to41600.java [hunk 1]
```diff
@@ -65,6 +70,28 @@ public boolean supportsRollingUpgrade() {
     @Override
     public void performDataMigration(Connection conn) {
         generateUuidForExistingSshKeyPairs(conn);
+        generateAnnotationPermissions(conn);
```
Comment (sureshanaparti) [inline engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41520to41600.java:73]:
```suggestion
       populateAnnotationPermissions(conn);
```

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41520to41600.java [hunk 2]
```diff
@@ -65,6 +70,28 @@ public boolean supportsRollingUpgrade() {
...
+    }
+
+    private void generateAnnotationPermissions(Connection conn) {
+        List<String> annotationRules = Arrays.asList("listAnnotations", "addAnnotation", "removeAnnotation");
+        for (RoleType roleType : Arrays.asList(RoleType.ResourceAdmin, RoleType.DomainAdmin, RoleType.User)) {
+            checkAnnotationPermissionExists(conn, roleType, annotationRules);
+        }
+    }
+
+    private void checkAnnotationPermissionExists(Connection conn, RoleType roleType, List<String> rules) {
```
Comment (sureshanaparti) [inline engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41520to41600.java:83]:
```suggestion
    private void checkAndPersistAnnotationPermissions(Connection conn, RoleType roleType, List<String> rules) {
```

**Comment by nvazquez** [comment]
@blueorangutan package

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41520to41600.java [hunk 3]
```diff
@@ -65,6 +70,28 @@ public boolean supportsRollingUpgrade() {
     @Override
     public void performDataMigration(Connection conn) {
         generateUuidForExistingSshKeyPairs(conn);
+        populateAnnotationPermissions(conn);
+    }
+
+    private void generateAnnotationPermissions(Connection conn) {
```
Comment (sureshanaparti) [inline engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41520to41600.java:76]:
```suggestion
    private void populateAnnotationPermissions(Connection conn) {
```

File: engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41520to41600.java [hunk 4]
```diff
@@ -65,6 +70,28 @@ public boolean supportsRollingUpgrade() {
...
+        List<String> annotationRules = Arrays.asList("listAnnotations", "addAnnotation", "removeAnnotation");
+        for (RoleType roleType : Arrays.asList(RoleType.ResourceAdmin, RoleType.DomainAdmin, RoleType.User)) {
+            checkAnnotationPermissionExists(conn, roleType, annotationRules);
```
Comment (sureshanaparti) [inline engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41520to41600.java:79]:
```suggestion
            checkAndPersistAnnotationPermissions(conn, roleType, annotationRules);
```

**Comment by nvazquez** [comment]
@blueorangutan package

**Review by sureshanaparti — APPROVED** [review]
code LGTM

**Comment by nvazquez** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]
clgtm,
@nvazquez I know you did some upgrade testing, can you describe that, please?

It seems to me we can use a similar mech for the guest os (mapping) upgrades in the future. Do you agree @nvazquez @sureshanaparti ? see #3609, #5685

**Comment by nvazquez** [comment]
@DaanHoogland I have added it to the PR description: tested the vanilla environment creation and also upgrade from a 4.15.2 environment in which I didn't add any additional role permission

**Comment by DaanHoogland** [comment]
@nvazquez @sureshanaparti I'd say one more test and we are good to merge:
4.15.2 env *with* extra roles/permissions defined -> upgrade it.
??

**Comment by sureshanaparti** [comment]
> @nvazquez @sureshanaparti I'd say one more test and we are good to merge:
> 4.15.2 env _with_ extra roles/permissions defined -> upgrade it.
> ??

@DaanHoogland I'll perform upgrade test with some rules, and update here.

**Comment by DaanHoogland** [comment]
> > @nvazquez @sureshanaparti I'd say one more test and we are good to merge:
> > 4.15.2 env _with_ extra roles/permissions defined -> upgrade it.
> > ??
> 
> @DaanHoogland I'll perform upgrade test with some rules, and update here.

thanks @sureshanaparti

**Comment by sureshanaparti** [comment]
Verified upgrade from CS 4.15.2 with existing annotations permissions. LGTM.

- Manually added the following rules to default roles (Resource Admin: 2, Domain Admin: 3, User: 4) in DB, as add/update rules to the default roles is not allowed from 4.15.

```
INSERT INTO `cloud`.`role_permissions` (uuid, role_id, rule, permission) VALUES (UUID(), 2, 'listAnnotations', 'ALLOW');
INSERT INTO `cloud`.`role_permissions` (uuid, role_id, rule, permission) VALUES (UUID(), 3, 'addAnnotation', 'DENY');
INSERT INTO `cloud`.`role_permissions` (uuid, role_id, rule, permission) VALUES (UUID(), 3, 'removeAnnotation', 'ALLOW');
INSERT INTO `cloud`.`role_permissions` (uuid, role_id, rule, permission) VALUES (UUID(), 4, 'listAnnotations', 'ALLOW');
INSERT INTO `cloud`.`role_permissions` (uuid, role_id, rule, permission) VALUES (UUID(), 4, 'addAnnotation', 'DENY');
INSERT INTO `cloud`.`role_permissions` (uuid, role_id, rule, permission) VALUES (UUID(), 4, 'removeAnnotation', 'DENY');
```

- Upgraded to 4.16, and noticed the missing rules are added. If there are any existing rules, this is the safe mechanism - to check and add the rules while upgrading.

```
2022-01-17 21:53:59,522 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Generating uuid for existing ssh key-pairs
2022-01-17 21:53:59,523 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Successfully generated uuid for existing ssh key-pairs
2022-01-17 21:53:59,525 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Checking the annotation permissions for the role: 2
2022-01-17 21:53:59,528 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Checking the annotation permissions for the role: 2 and rule: listAnnotations
2022-01-17 21:53:59,531 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Found existing role permission for role: 2 and rule: listAnnotations, not updating it
2022-01-17 21:53:59,531 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Checking the annotation permissions for the role: 2 and rule: addAnnotation
2022-01-17 21:53:59,533 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Inserting role permission for role: 2 and rule: addAnnotation
2022-01-17 21:53:59,534 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Checking the annotation permissions for the role: 2 and rule: removeAnnotation
2022-01-17 21:53:59,534 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Inserting role permission for role: 2 and rule: removeAnnotation
2022-01-17 21:53:59,535 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Checking the annotation permissions for the role: 3
2022-01-17 21:53:59,535 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Checking the annotation permissions for the role: 3 and rule: listAnnotations
2022-01-17 21:53:59,535 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Inserting role permission for role: 3 and rule: listAnnotations
2022-01-17 21:53:59,536 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Checking the annotation permissions for the role: 3 and rule: addAnnotation
2022-01-17 21:53:59,536 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Found existing role permission for role: 3 and rule: addAnnotation, not updating it
2022-01-17 21:53:59,536 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Checking the annotation permissions for the role: 3 and rule: removeAnnotation
2022-01-17 21:53:59,537 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Found existing role permission for role: 3 and rule: removeAnnotation, not updating it
2022-01-17 21:53:59,537 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Checking the annotation permissions for the role: 4
2022-01-17 21:53:59,537 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Checking the annotation permissions for the role: 4 and rule: listAnnotations
2022-01-17 21:53:59,538 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Found existing role permission for role: 4 and rule: listAnnotations, not updating it
2022-01-17 21:53:59,538 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Checking the annotation permissions for the role: 4 and rule: addAnnotation
2022-01-17 21:53:59,538 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Found existing role permission for role: 4 and rule: addAnnotation, not updating it
2022-01-17 21:53:59,538 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Checking the annotation permissions for the role: 4 and rule: removeAnnotation
2022-01-17 21:53:59,539 DEBUG [c.c.u.d.Upgrade41520to41600] (main:null) (logid:) Found existing role permission for role: 4 and rule: removeAnnotation, not updating it
2022-01-17 21:53:59,553 INFO  [c.c.u.DatabaseUpgradeChecker] (main:null) (logid:) Cleanup upgrade Upgrade41520to41600 to upgrade from 4.15.2.0-4.16.0.0 to 4.16.0.0
```

**Pull request merged**